### PR TITLE
[CRDT] FT.DROPINDEX FT.SUG

### PR DESCRIPTION
**Pull request by ashtul**

### CODE REVIEW

File: src/module.c [hunk 1]
```diff
@@ -413,7 +413,8 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
     RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
   } else {
-    RedisModule_Replicate(ctx, RS_DROP_INDEX_IF_X_CMD, "v", argv + 1, argc - 1);
+    // Remove DD if CRDT. Documents were deleted with RM_Call.
+    RedisModule_Replicate(ctx, RS_DROP_INDEX_IF_X_CMD, "v", argv + 1, argc - 1 - (!!isCrdt & delDocs));
```
Comment (rafie) [inline src/module.c:417]: &&

Comment (ashtul) [reply src/module.c:417]: Will it equal `1`? We neglect the `DD` flag this way.

File: src/module.c [hunk 2]
```diff
@@ -413,7 +413,8 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
     RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
```
Comment (MeirShpilraien) [inline src/module.c:414]: I think we should handle it here as well

File: src/module.c (hunk 1 above)
Comment (MeirShpilraien) [reply src/module.c:417]: Shouldn't it be: (!!(isCrdt && delDocs))?

File: src/redis_index.c [hunk 3]
```diff
@@ -462,6 +462,20 @@ int Redis_DeleteKey(RedisModuleCtx *ctx, RedisModuleString *s) {
   return res;
 }
 
+int Redis_DeleteKeyC(RedisModuleCtx *ctx, char *cstr, int replicate) {
+  RedisModuleCallReply *rep;
+  if (!replicate) {
+    rep = RedisModule_Call(ctx, "DEL", "c", cstr);
+  } else {
+    // for CRDT we need to send command to all replicas. 
+    rep = RedisModule_Call(ctx, "DEL", "c!", cstr);
```
Comment (MeirShpilraien) [inline src/redis_index.c:471]: We need to verify if the ctx used here is the client ctx so it will automatically be wrapt with multi exec.

**Review by MeirShpilraien — COMMENTED** [review]
Some small comments and things to verify

File: src/module.c (hunk 2 above)
Comment (ashtul) [reply src/module.c:414]: @MeirShpilraien  `FT.DROP` is deprecated on RS 2.0. Should we fix it anyway?

File: src/redis_index.c (hunk 3 above)
Comment (ashtul) [reply src/redis_index.c:471]: The `ctx` is the one we get with `FT.DROPINDEX`. Am I missing something?

File: src/module.c (hunk 2 above)
Comment (MeirShpilraien) [reply src/module.c:414]: Sorry you are right :+1:

Comment (MeirShpilraien) [reply src/module.c:414]: Although I do think we should fix this as well, we do not want someone to do it and crash the server..

File: src/redis_index.c (hunk 3 above)
Comment (MeirShpilraien) [reply src/redis_index.c:471]: If its the same ctx we get with the command then we are good.

File: src/module.c (hunk 2 above)
Comment (ashtul) [reply src/module.c]: Done

**Comment by ashtul** [comment]
Tests are at RedisLabsModules/RediSearch-Enterprise-Tests#5

File: src/module.c [hunk 4]
```diff
@@ -411,9 +411,18 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 
   if (RMUtil_StringEqualsCaseC(argv[0], "FT.DROP") ||
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
-    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
+    if (isCrdt) {
+      // We always send KEEPDOC. Either the user specified it or we deleted it with RM_Call
+      RedisModuleString *keepDocStr = RedisModule_CreateString(ctx, "KEEPDOCS", strlen("KEEPDOCS"));
+      RedisModuleString *repCmd[2] = { argv[1], keepDocStr };
+      RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", keepDocStr, 2);
```
Comment (MeirShpilraien) [inline src/module.c:418]:
```suggestion
      RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", repCmd, 2);
```

**Review by MeirShpilraien — COMMENTED** [review]
Small fix (or maybe I miss something)

File: src/module.c [hunk 5]
```diff
@@ -412,9 +412,14 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 
   if (RMUtil_StringEqualsCaseC(argv[0], "FT.DROP") ||
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
-    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
+    // We always send KEEPDOC to the slave.
+    RedisModuleString *keepDocStr = RedisModule_CreateString(ctx, "KEEPDOCS", strlen("KEEPDOCS"));
+    RedisModuleString *repCmd[2] = { argv[1], keepDocStr };
+    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", repCmd, 2);
```
Comment (MeirShpilraien) [inline src/module.c:418]:
```suggestion
    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "sc", argv[1], "KEEPDOCS");
```

File: src/module.c [hunk 6]
```diff
@@ -412,9 +412,14 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 
   if (RMUtil_StringEqualsCaseC(argv[0], "FT.DROP") ||
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
-    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
+    // We always send KEEPDOC to the slave.
+    RedisModuleString *keepDocStr = RedisModule_CreateString(ctx, "KEEPDOCS", strlen("KEEPDOCS"));
+    RedisModuleString *repCmd[2] = { argv[1], keepDocStr };
```
Comment (MeirShpilraien) [inline src/module.c:417]:
```suggestion

```

File: src/module.c [hunk 7]
```diff
@@ -412,9 +412,14 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 
   if (RMUtil_StringEqualsCaseC(argv[0], "FT.DROP") ||
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
-    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
+    // We always send KEEPDOC to the slave.
+    RedisModuleString *keepDocStr = RedisModule_CreateString(ctx, "KEEPDOCS", strlen("KEEPDOCS"));
```
Comment (MeirShpilraien) [inline src/module.c:416]:
```suggestion

```

File: src/module.c [hunk 8]
```diff
@@ -412,9 +412,14 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
+    RedisModuleString *repCmd[2] = { argv[1], keepDocStr };
+    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", repCmd, 2);
+    RedisModule_FreeString(ctx, keepDocStr);
```
Comment (MeirShpilraien) [inline src/module.c:419]:
```suggestion

```

File: src/redis_index.c [hunk 9]
```diff
@@ -462,6 +462,15 @@ int Redis_DeleteKey(RedisModuleCtx *ctx, RedisModuleString *s) {
   return res;
 }
 
+int Redis_DeleteKeyC(RedisModuleCtx *ctx, char *cstr) {
+  RedisModuleCallReply *rep;
+  rep = RedisModule_Call(ctx, "DEL", "c!", cstr);
```
Comment (MeirShpilraien) [inline src/redis_index.c:467]: We need to replicate (`!`) only if its not crdt

**Review by MeirShpilraien — COMMENTED** [review]
Small comments and suggestion, looks good!! :+1:

**Review by MeirShpilraien — APPROVED** [review]
:+1:

File: src/module.c [hunk 10]
```diff
@@ -412,9 +412,11 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 
   if (RMUtil_StringEqualsCaseC(argv[0], "FT.DROP") ||
       RMUtil_StringEqualsCaseC(argv[0], "_FT.DROP")) {
-    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "v", argv + 1, argc - 1);
+    // We always send KEEPDOC to the slave.
+    RedisModule_Replicate(ctx, RS_DROP_IF_X_CMD, "sc", argv[1], "KEEPDOCS");
   } else {
-    RedisModule_Replicate(ctx, RS_DROP_INDEX_IF_X_CMD, "v", argv + 1, argc - 1);
+    // Remove DD as documents were deleted with RM_Call.
+    RedisModule_Replicate(ctx, RS_DROP_INDEX_IF_X_CMD, "v", argv + 1, argc - 1 - !!delDocs);
```
Comment (rafie) [inline src/module.c:419]: `!!delDocs` -> `(delDocs ? 1 : 0)` or even just `delDocs` (it should be either 1 or 0, right?)

**Review by rafie — APPROVED** [review]

**Pull request merged**